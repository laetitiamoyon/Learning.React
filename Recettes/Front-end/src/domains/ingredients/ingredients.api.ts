import axios, { AxiosResponse, AxiosRequestConfig } from 'axios'
import { Ingredient } from './ingredients.model';

export const url : string = 'http://localhost:3001/ingredients'
export const configuration : AxiosRequestConfig = { headers : { 'Content-Type' : 'application/json' } }

export const getAllIngredients = async () : Promise<AxiosResponse<Ingredient[]>> =>
    (await axios.get(url, configuration)).data

export const addIngredient = async (ingredient : Ingredient) : Promise<AxiosResponse<Ingredient>> =>
    (await axios.post(url, ingredient, configuration)).data

export const updateIngredient = async (ingredient: Ingredient) : Promise<AxiosResponse<Ingredient>> =>
    (await axios.put(`${url}/${ingredient.id}`, ingredient, configuration)).data

export const removeIngredient = async (id : string) : Promise<AxiosResponse<Ingredient>> =>
    (await axios.delete(`${url}/${id}`, configuration)).data