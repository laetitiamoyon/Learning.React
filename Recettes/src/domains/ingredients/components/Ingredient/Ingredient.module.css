
.formContainer
{
    margin-top: 50px;

    display: flex;
    justify-content: center;

}

.ingredientTitle, .ingredientQuantity, .ingredientUnity
{
    width: 130px;

    border: #BDBDBD solid 1px;
    border-radius: 8px;
    color: #394044;

    text-align: center;

    margin-left: 20px;
}

.title { margin-top: 20px;}


.removeButton 
{
    background-image: url(./Icons/trash.png);
    background-size: cover;
    width: 25px;
    height: 25px;

    cursor: pointer;

    margin-left: 20px;
}

.updateButton
{
    width: 130px;
    height: 30px;

    line-height: 26px;
    text-align: center;
    color: #394044;
    font-weight: 600;

    background-color: #F1EEEB;

    cursor: pointer;

    border: #BDBDBD solid 1px;
    border-radius: 8px;

    margin-left: 20px;
}
.updateButton:hover {

    background-color: #ffae71;
    border:  #ffae71;
    color: white;
}

