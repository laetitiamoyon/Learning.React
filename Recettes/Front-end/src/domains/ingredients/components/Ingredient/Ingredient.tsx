import { FC, useState, ChangeEvent } from 'react';
import { Ingredient as IngredientModel } from '../../ingredients.model';
import styles from './Ingredient.module.scss'
import { useDispatch } from 'react-redux';
import { removeIngredientRequestAction, editIngredientRequestAction } from '../../ingredients.actions';
import { updateRecipeIngredientRequestAction, removeRecipeIngredientRequestAction } from '../../../recipes/recipes.actions';
import { Highlight } from '../../../../shared/utils/Highlight';

type CurrentWindow = 
    'EDIT_INGREDIENT_BUTTON' |
    'UPDATE_INGREDIENT'

interface Props extends IngredientModel
{
    searchIngredientTerm : string,
    color : string
}

const Ingredient :FC<Props> = ({ id, title, unity, searchIngredientTerm, color }) =>
{
  const [newTitle, setNewTitle] = useState(title)
  const [newUnity, setNewUnity] = useState(unity)
  const [currentWindow, setWindow] = useState<CurrentWindow>('EDIT_INGREDIENT_BUTTON')

  const clickOnEditIngredient = () : void => setWindow('UPDATE_INGREDIENT')

  const onChangeTitle = (event : ChangeEvent<HTMLInputElement>) => setNewTitle(event.target.value)
  const onChangeUnity = (event : ChangeEvent<HTMLInputElement>) => setNewUnity(event.target.value)

  const ingredientDispatch = useDispatch()
  const recipeDispatch = useDispatch()

  const removeIngredient = () => 
  {
    ingredientDispatch(removeIngredientRequestAction(id))
    recipeDispatch(removeRecipeIngredientRequestAction(id))
  }

  const updateRecipeIngredient = () => 
  {
    ingredientDispatch(editIngredientRequestAction(
    { 
      id, 
      title : newTitle,
      unity : newUnity
    }))
    recipeDispatch(updateRecipeIngredientRequestAction({
      id,
      title : newTitle,
      unity : newUnity
    }))
  }

  return <div className={styles.container}>
    <div className={styles.ingredient}>
    { currentWindow === 'EDIT_INGREDIENT_BUTTON' && 
            <div onClick={clickOnEditIngredient} className={styles.title}>
                <Highlight searchTerm={searchIngredientTerm} text={title} color={color}/></div>}
        
            { currentWindow === 'UPDATE_INGREDIENT' &&
                <input
                    className={styles.input} 
                    placeholder="titre"
                    onChange={onChangeTitle} 
                    value={newTitle}/> 
            }

      <input 
        className={styles.input} 
        value={newUnity}
        onChange={onChangeUnity}/>

      <button className={styles.updateButton} onClick={updateRecipeIngredient}>Mettre Ã  jour</button>
      <div className={styles.trashIcon} onClick={removeIngredient}></div>
    </div>
  </div>
};

export default Ingredient;