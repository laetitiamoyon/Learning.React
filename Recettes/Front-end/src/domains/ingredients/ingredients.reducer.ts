import { failedActionMetadata, loadedActionMetadata, loadingActionMetadata } from '../../shared/domains/Redux/redux.utils';
import { IngredientAction, IngredientActions } from './ingredients.actions';
import { IngredientState, initialIngredients } from './ingredients.state';


export const ingredientsReducer = (state: IngredientState = initialIngredients , action: IngredientActions): IngredientState =>
{
    switch (action.type)
    {
        case IngredientAction.GET_INGREDIENTS_REQUEST : return loadingActionMetadata(state)
        case IngredientAction.GET_INGREDIENTS_SUCCESS : return {
            ...loadedActionMetadata(state),
            ingredients: action.payload
        }
        case IngredientAction.GET_INGREDIENTS_FAILED : return failedActionMetadata(action.error, state)
  
        default : return state
    }
}