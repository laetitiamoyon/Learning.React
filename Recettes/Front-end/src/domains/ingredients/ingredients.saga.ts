import { call, put, takeLatest } from 'redux-saga/effects'
import { addIngredientFailedAction, AddIngredientRequestAction, addIngredientSuccessAction, editIngredientFailedAction, EditIngredientRequestAction, editIngredientSuccessAction, getIngredientsFailedAction, getIngredientsRequestAction, GetIngredientsRequestAction, getIngredientsSuccessAction, IngredientAction, removeIngredientFailedAction, RemoveIngredientRequestAction, RemoveIngredientSuccessAction } from './ingredients.actions';
import { Ingredient } from './ingredients.model';
import { addIngredient, getAllIngredients, updateIngredient, removeIngredient } from './ingredients.api';

export function* getAllIngredientsSaga()
{
   try
   {
      const ingredients : Ingredient[] = yield call(getAllIngredients);
  
      yield put(getIngredientsSuccessAction(ingredients));
   }
   catch (error : any)
   {
      yield put(getIngredientsFailedAction(error.message));
   }
}

export function* addIngredientSaga(action : AddIngredientRequestAction)
{
   try
   {
      const newIngredient : Ingredient = yield call(addIngredient, action.payload);
      
      yield put(addIngredientSuccessAction(newIngredient));
      yield put(getIngredientsRequestAction());
   }
   catch (error : any)
   {
      yield put(addIngredientFailedAction(error.message));
   }
}

export function* editIngredientSaga(action : EditIngredientRequestAction)
{
   try
   {
      const ingredient = action.payload
      const editedIngredient : Ingredient = yield call(updateIngredient, ingredient);
     
      yield put(editIngredientSuccessAction(editedIngredient));
      yield put(getIngredientsRequestAction());
   }
   catch (error : any)
   {
      yield put(editIngredientFailedAction(error.message));
   }
}

export function* removeIngredientSaga(action : RemoveIngredientRequestAction)
{
   try
   {
        const { id } = action.payload

        yield call(removeIngredient, id);
        yield put(RemoveIngredientSuccessAction(id));
        yield put(getIngredientsRequestAction());
      }
   catch (error : any)
   {
      yield put(removeIngredientFailedAction(error.message));
   }
}

export default function* ingredientSaga() : Generator {
  yield takeLatest(IngredientAction.GET_INGREDIENTS_REQUEST, getAllIngredientsSaga);
  yield takeLatest(IngredientAction.ADD_INGREDIENT_REQUEST, addIngredientSaga);
  yield takeLatest(IngredientAction.EDIT_INGREDIENT_REQUEST, editIngredientSaga);
  yield takeLatest(IngredientAction.REMOVE_INGREDIENT_REQUEST, removeIngredientSaga);
}
