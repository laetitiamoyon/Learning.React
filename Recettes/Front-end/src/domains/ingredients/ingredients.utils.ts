import { Ingredient } from "./ingredients.model";

export const filterIngredients = (ingredients : Ingredient[], searchTerm : string) : Ingredient[] =>
    ingredients.filter(r => r.title.toLowerCase().includes(searchTerm.toLowerCase()))

// TODO : tester la couche utils (__tests__/ingredients.utils.test.ts)