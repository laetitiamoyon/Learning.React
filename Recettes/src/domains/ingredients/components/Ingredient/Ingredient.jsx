import React, { useState, useContext } from 'react';
import { IngredientContext } from '../../ingredients.provider';
import styles from './Ingredient.module.scss'

const Ingredient = ({id, title, quantity, unity}) => 
{
    const [newTitle, setNewTitle] = useState(title)
    const [newQuantity, setNewQuantity] = useState(quantity)
    const [newUnity, setNewUnity] = useState(unity)

    const { dispatch } = useContext(IngredientContext)
    const removeIngredient = () => dispatch({ type: 'REMOVE_INGREDIENT', payload : { id } })
    const updateIngredient = () => dispatch(
    { 
        type: 'UPDATE_INGREDIENT',
        payload :
        { 
            id : parseInt(id), 
            title : newTitle,
            unity : newUnity,
            quantity : parseInt(newQuantity),
        } 
    })

    const onChangeTitle = (event) => setNewTitle(event.target.value)
    const onChangeQuantity = (event) => setNewQuantity(event.target.value)
    const onChangeUnity = (event) => setNewUnity(event.target.value)

    console.log(unity)
    return <form className={styles.formContainer}>
        <input 
            placeholder="titre"
            onChange={onChangeTitle}
            value={newTitle}/>

        <input 
            placeholder="unité"
            onChange={onChangeUnity}
            value={newUnity}/>

        <div className={styles.updateButton} onClick={updateIngredient}>Mettre à jour</div>
        <div className={styles.removeButton} onClick={removeIngredient}></div>
    </form>
};

export default Ingredient;