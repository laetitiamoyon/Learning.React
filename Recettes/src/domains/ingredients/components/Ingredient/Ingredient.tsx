import { useState, useContext, ChangeEvent, FC } from 'react';
import { removeIngredientAction, updateIngredientAction } from '../../ingredients.action';
import { IngredientContext } from '../../ingredients.context';
import { Ingredient as IngredientModel } from '../../ingredients.model';
import styles from './Ingredient.module.scss'
import { RecipeContext } from '../../../recipes/recipes.context';
import { removeIngredientRecipeAction, updateIngredientRecipeAction } from '../../../recipes/recipes.action';

const Ingredient : FC<IngredientModel> = ({id, title, unity}) => 
{
    const [newTitle, setNewTitle] = useState(title)
    const [newUnity, setNewUnity] = useState(unity)

    const { dispatch: ingredientDispatch } = useContext(IngredientContext)
    const { dispatch : recipeDispatch } = useContext(RecipeContext)

    const removeIngredient = () => ingredientDispatch(removeIngredientAction(id))
    const removeRecipeIngredient = () => recipeDispatch(removeIngredientRecipeAction(id))
    
    const updateRecipeIngredient = () => {
        const ingredient : IngredientModel = {
            id, 
            title : newTitle,
            unity : newUnity
        }
        
        recipeDispatch(updateIngredientRecipeAction(ingredient))
    }

    const updateIngredient = () => ingredientDispatch(updateIngredientAction(
    { 
        id, 
        title : newTitle,
        unity : newUnity
    }))

    const onChangeTitle = (event: ChangeEvent<HTMLInputElement>) : void => setNewTitle(event.target.value)
    const onChangeUnity = (event: ChangeEvent<HTMLInputElement>) : void => setNewUnity(event.target.value)

    return <form className={styles.formContainer}>
        <input 
            placeholder="titre"
            onChange={onChangeTitle}
            value={newTitle}/>

        <input 
            placeholder="unité"
            onChange={onChangeUnity}
            value={newUnity}/>

        <div className={styles.updateButton} onClick={() => {
            updateIngredient();
            updateRecipeIngredient();}}>Mettre à jour</div>
        <div className={styles.removeButton} onClick={() => {
            removeIngredient();
            removeRecipeIngredient();}}></div>
    </form>
};

export default Ingredient;