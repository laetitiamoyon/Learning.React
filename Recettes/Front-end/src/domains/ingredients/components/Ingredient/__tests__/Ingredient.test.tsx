import {fireEvent, getByText, render} from "@testing-library/react";
import {useDispatch} from "react-redux";
import Ingredient from "../Ingredient";
import {ingredientMock} from "../../../ingredient.mock";
import {editIngredientRequestAction, removeIngredientRequestAction} from "../../../ingredients.actions";
import {
    removeRecipeIngredientRequestAction,
    updateRecipeIngredientRequestAction
} from "../../../../recipes/recipes.actions";

jest.mock('react-redux')
const mockUseDispatch = useDispatch as jest.MockedFunction<typeof useDispatch>
const mockDispatch = jest.fn()

describe("Ingredient", () => {

    beforeEach(() => (mockUseDispatch as jest.Mock).mockReturnValue(mockDispatch))

    it("should match snapshot", () => {
        // Given & When
        const {container} = render(<Ingredient {...ingredientMock} color={""} searchIngredientTerm={""}/>)

        // Then
        expect(container).toMatchSnapshot()
    })

    it("should dispatch editIngredientRequestAction and updateRecipeIngredientRequestAction when add updateRecipeIngredient button is clicked", () =>
    {
        // Given
        const {getByText} = render(<Ingredient {...ingredientMock} color={""} searchIngredientTerm={""}/>)
        const button = getByText("Mettre à jour")

        // When
        fireEvent.click(button)

        // Then
        expect(mockDispatch).toBeCalledWith(editIngredientRequestAction(ingredientMock));
        expect(mockDispatch).toBeCalledWith(updateRecipeIngredientRequestAction(ingredientMock));
    })

    it("should dispatch removeIngredientRequestAction and removeRecipeIngredientRequestAction when trash icon is clicked", () =>
    {
        // Given
        const {container} = render(<Ingredient {...ingredientMock} color={""} searchIngredientTerm={""}/>)
        const trashIcon = container.querySelector('button + div')!

        // When
        fireEvent.click(trashIcon)

        // Then
        expect(mockDispatch).toBeCalledWith(removeIngredientRequestAction(ingredientMock.id))
        expect(mockDispatch).toBeCalledWith(removeRecipeIngredientRequestAction(ingredientMock.id));
    })
})

// Tester que les éléments attendus sont présent quand la current window change.
// Quand le state ou le searchterm est modifiée, vérifier que filterIngredients est appelé avec les bons paramètres et renvoit les bons paramètres