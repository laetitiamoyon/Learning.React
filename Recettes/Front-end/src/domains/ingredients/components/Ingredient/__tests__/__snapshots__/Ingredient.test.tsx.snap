// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingredient should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="ingredient"
    >
      <div
        class="title"
      >
        <div>
          <span
            style="color :"
          />
          P
          <span
            style="color :"
          />
          â
          <span
            style="color :"
          />
          t
          <span
            style="color :"
          />
          e
          <span
            style="color :"
          />
           
          <span
            style="color :"
          />
          à
          <span
            style="color :"
          />
           
          <span
            style="color :"
          />
          p
          <span
            style="color :"
          />
          i
          <span
            style="color :"
          />
          z
          <span
            style="color :"
          />
          z
          <span
            style="color :"
          />
          a
          <span
            style="color :"
          />
        </div>
      </div>
      <input
        class="input"
        value="pièce"
      />
      <button
        class="updateButton"
      >
        Mettre à jour
      </button>
      <div
        class="trashIcon"
      />
    </div>
  </div>
</div>
`;
